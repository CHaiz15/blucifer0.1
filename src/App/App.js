import React, { Component } from 'react';
import './App.css';
import { getData, fetchAreaDetails, fetchListings } from '../apiCalls/apiCalls';
import LoginForm from '../LoginForm/LoginForm.js';
import Header from '../Header/Header.js';
import AreasContainer from '../AreasContainer/AreasContainer.js';
import ListingsContainer from '../ListingsContainer/ListingsContainer.js';
import ListingDetails from '../ListingDetails/ListingDetails.js';
import Favorites from '../Favorites/Favorites.js';
import { Route } from 'react-router-dom';

class App extends Component {
  constructor() {
    super();
    this.state = {
      areas: [],
      listings: [],
      selectedAreaId: '',
      selectedListingId: '',
      user: {
        name: '',
        purpose: ''
      },
      favoriteListings: []
    }
  }

  addFormInfo = (info) => {
    this.setState({user: info})
  }
  
  addSelectedArea = (id) => {
    this.setState({selectedAreaId: id})
  } 

  addSelectedListing = (id) => {
    this.setState({selectedListingId: id})
  }

  addFavoritedListing = (id) => {
    !this.state.favoriteListings.includes(id) && this.setState({favoriteListings: [...this.state.favoriteListings, id]})
  } 

  removeFavoritedListing = (id) => {
    const filteredListings = this.state.favoriteListings.filter(listing => {
      return listing !== id
    })
    this.setState({favoriteListings: filteredListings})
  }

  componentDidMount() {
    getData()
      .then(areaData => fetchAreaDetails(areaData))
      .then(areas => {
        this.setState({areas: areas})
        const arrOfSeperateListings = areas.map(area => area.listings)
        return arrOfSeperateListings.reduce((acc, listingArray) => acc.concat(listingArray), [])
      })
      .then(listings => fetchListings(listings))
      .then(listings => this.setState({listings: listings}))
  }

  render() {
    return (
      <main>
        <Route exact path='/' render={() => <LoginForm addFormInfo={this.addFormInfo}/>} />
        <Route path='/nav' render={() => 
          <Header 
            name={this.state.user.name} 
            purpose={this.state.user.purpose}
            totalFavorites={this.state.favoriteListings.length}
          />} 
        />
        <Route exact path='/nav/areas' render={() => 
          <AreasContainer 
            addSelectedArea={this.addSelectedArea} 
            areas={this.state.areas}
          />} 
        />
        <Route exact path='/nav/areas/:area_id/listings' render={() => 
          <ListingsContainer 
            addSelectedListing={this.addSelectedListing} 
            selectedAreaId={this.state.selectedAreaId} 
            listings={this.state.listings} 
            areas={this.state.areas}
           />} 
        />
        <Route path='/nav/favorites' render={() => 
          <Favorites 
            addSelectedListing={this.addSelectedListing} 
            selectedAreaId={this.state.selectedAreaId} 
            listings={this.state.listings} 
            favoriteListings={this.state.favoriteListings}
          />} 
        />
        <Route exact path='/nav/areas/:area_id/listings/:listing_id' render={() => 
          <ListingDetails 
            key={this.state.selectedListingId} 
            addFavoritedListing={this.addFavoritedListing}
            removeFavoritedListing={this.removeFavoritedListing} 
            selectedListingId={this.state.selectedListingId} 
            listings={this.state.listings} 
            favoriteListings={this.state.favoriteListings}
          />} 
        /> 
      </main>
    )
  }
}
export default App;
