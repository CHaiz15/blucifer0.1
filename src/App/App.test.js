import React from 'react';
import App from './App';
import { shallow } from 'enzyme'
import { getData, fetchAreaDetails, fetchListings } from '../apiCalls/apiCalls'

jest.mock('../apiCalls/apiCalls.js')

describe('App', () => {
  let wrapper;
  beforeEach(() => {
    getData.mockImplementation(() => {
      return Promise.resolve({ areas: [
        {
          area: "RiNo",
          details: '/api/v1/areas/590'
        }
      ]})
    });

    fetchAreaDetails.mockImplementation(() => {
      return Promise.resolve({ areas: [
        {
          id: 1,
          shortName: 'RiNo',
          name: 'Rhino',
          location: 'Denver',
          about: 'Cool place',
          listings: '/api/v1/listings/3'
        }
      ]})
    });

    fetchListings.mockImplementation(() => {
      return Promise.resolve({ areas: [
        {
          listing_id: 3,
          area_id: 590,
          name: 'RiNo',
          address: '123 Happy street',
          details: []
        }
      ]})
    });
  });

  it('should be an instance of App component', () => {
    wrapper = shallow(<App/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should retrieve areas when mounting', () => {
    wrapper = shallow(<App />);
    expect(getData).toHaveBeenCalled();
  });

  it('should retrieve area details when mounting', () => {
    wrapper = shallow(<App />);
    expect(fetchAreaDetails).toHaveBeenCalled();
  });

  it('should retrieve listings when mounting', () => {
    wrapper = shallow(<App />);
    expect(fetchAreaDetails).toHaveBeenCalled();
  });

  it('it should have default state', () => {
    wrapper = shallow(<App />);
    expect(wrapper.state('areas')).toEqual([]);
    expect(wrapper.state('listings')).toEqual([]);
    expect(wrapper.state('selectedAreaId')).toEqual('');
    expect(wrapper.state('selectedListingId')).toEqual('');
    expect(wrapper.state('user')).toEqual({name: '', purpose: ''});
    expect(wrapper.state('favoriteListings')).toEqual([]);
  });

  it('should update user when addFormInfo is called', () => {
    const mockFormInfo = {name: 'John Doe', purpose: 'Business'};
    wrapper.instance().addFormInfo(mockFormInfo);
    expect(wrapper.state('user')).toEqual({name: 'John Doe', purpose: 'Business'});
  });

  it('should update selectedAreaId when addSelectedArea is called', () => {
    const mockSelectedAreaId = 590;
    wrapper.instance().addSelectedArea(mockSelectedAreaId);
    expect(wrapper.state('selectedAreaId')).toEqual(590);
  });

  it('should update selectedListingId when addSelectedListing is called', () => {
    const mockSelectedListingId = 3;
    wrapper.instance().addSelectedListing(mockSelectedListingId);
    expect(wrapper.state('selectedListingId')).toEqual(3);
  });

  it('should add a favoriteListingId to favoriteListings when addFavoritedListing is called', () => {
    const mockfavoriteListingId = 3;
    wrapper.instance().addFavoritedListing(mockfavoriteListingId);
    expect(wrapper.state('favoriteListings')).toEqual([3]);
  });

  it('should remove a favoriteListingId from favoriteListings when removeFavoritedListing is called', () => {
    const mockremovefavoriteListingId = 3;
    expect(wrapper.state('favoriteListings')).toEqual([3]);
    wrapper.instance().removeFavoritedListing(mockremovefavoriteListingId);
    expect(wrapper.state('favoriteListings')).toEqual([]);
  });
})