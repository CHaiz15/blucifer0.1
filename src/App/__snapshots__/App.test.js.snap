// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should be an instance of App component 1`] = `
<main>
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    path="/nav"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/nav/areas"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/nav/areas/:area_id/listings"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/nav/areas/:area_id/listings/:listing_id"
    render={[Function]}
  />
</main>
`;
